### ISSUE TITLE
No option to force overwrite changes

### ISSUE THREAD

**Issue by MG2R** [puppetlabs/r10k#483]
When running r10k on a filesystem with local changes, R10K errors out with

```
ERROR    -> Command exited with non-zero exit code:
Command: git --git-dir /etc/puppetlabs/puppet/environments/SYS1042/modules/yum_config/.git --work-tree /etc/puppetlabs/puppet/environments/SYS1042/modules/yum_config checkout 5d2e4082f939536f97f3495c5ef93c444f52b70c
Stderr:
error: Your local changes to the following files would be overwritten by checkout:
    templates/yum.conf.erb
Please, commit your changes or stash them before you can switch branches.
Aborting
Exit code: 1
```

There should be an option to be more aggressive and force-overwrite the local changes with changes from upstream.

**Comment by kmcfate** [comment]
This also occurs when pinning to a tag or release. There is no need for local changes. Especially with changes around .gitignore. So, honestly I would consider this a bug, not an enhancement, since it breaks pinning for many repositories.

**Comment by cyberious** [comment]
Agreed this is clearly a bug, I am fighting with this one on a regular basis.

**Comment by alberto-sierra** [comment]
+1

**Comment by lasko** [comment]
+1

**Comment by scotje** [comment]
The flag still needs to be added (see #639) but can you guys test on 2.4.x? The default behavior has changed.

**Comment by scotje** [comment]
The JIRA ticket https://tickets.puppetlabs.com/browse/RK-264 now exists to track adding the flag.

**Comment by hdeadman** [comment]
I am using 2.4.3 and the default behaviour for r10k install is different than it used to be (overwite everything) but it is also detecting local modifications when there are none (so those modules aren't updated). It seems like all the modules that are checked out of git are deemed to have local modifications and they aren't overwritten or updated. If I do git status in a module that is reported to have local modifications it says:

# HEAD detached at a45a562

nothing to commit, working directory clean

This issue is labelled as enhancement, is there another issue tracking a bug?

**Comment by scotje** [comment]
I've got a PR up now which will add some debug logging when r10k detects local modifications:

https://github.com/puppetlabs/r10k/pull/657

I have not been able to reproduce `r10k puppetfile install` thinking that a module checkout has local modifications when it doesn't though. Once that PR is merged, perhaps testing with r10k/master will shed some light on the situation.

**Comment by sfhardman** [comment]
Not sure if this is the problem others are having with unexpected local modifications, here is where I'm hitting the problem:
puppet/
  .git/
  puppetfile
  modules/
    r10k_managed_git_module/
      .git/

I run a git clean -fdx on the puppet working directory after checkout, and this removes the files from r10k_managed_git_module directory _except_ for the r10k_managed_git_module/.git directory.

The result is a dirty r10k_managed_git_module working directory

**Comment by hdeadman** [comment]
@scotje I merged your PR 657 to a fork and ran r10k with debug turned on. In my case all the modules checked out from git were showing the warning that they are being skipped due to local modifications. I am running on redhat 7.2 with git version 1.8.3.1. 

This is the section of output from one of the modules that was "clean" but was being reported as having local modifications. 

```
[2016-10-14 21:26:02 - INFO] Updating module /vagrant/puppet/3rdparty/modules/solr
[2016-10-14 21:26:02 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "rev-parse", "master^{commit}"]
[2016-10-14 21:26:02 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git rev-parse master^{commit}
Stdout:
03c429c2ba8305e75bb23b9e9b5f5bec057e90dd
Exit code: 0
[2016-10-14 21:26:02 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "for-each-ref", "refs/heads", "--format", "%(refname)"]
[2016-10-14 21:26:02 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git for-each-ref refs/heads --format %(refname)
Stdout:
refs/heads/master
Exit code: 0
[2016-10-14 21:26:02 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "config", "--local", "--get-regexp", "^remote\\..*\\.url$"]
[2016-10-14 21:26:02 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git config --local --get-regexp ^remote\..*\.url$
Stdout:
remote.origin.url https://bitbucket.org/hdeadman/puppet-solr.git
Exit code: 0
[2016-10-14 21:26:02 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "fetch", "origin", "--prune"]
[2016-10-14 21:26:03 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git fetch origin --prune
Exit code: 0
[2016-10-14 21:26:03 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "rev-parse", "master^{commit}"]
[2016-10-14 21:26:03 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git rev-parse master^{commit}
Stdout:
03c429c2ba8305e75bb23b9e9b5f5bec057e90dd
Exit code: 0
[2016-10-14 21:26:03 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "rev-parse", "master^{commit}"]
[2016-10-14 21:26:03 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git rev-parse master^{commit}
Stdout:
03c429c2ba8305e75bb23b9e9b5f5bec057e90dd
Exit code: 0
[2016-10-14 21:26:03 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "for-each-ref", "refs/heads", "--format", "%(refname)"]
[2016-10-14 21:26:03 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git for-each-ref refs/heads --format %(refname)
Stdout:
refs/heads/master
Exit code: 0
[2016-10-14 21:26:03 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "rev-parse", "master^{commit}"]
[2016-10-14 21:26:03 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git rev-parse master^{commit}
Stdout:
03c429c2ba8305e75bb23b9e9b5f5bec057e90dd
Exit code: 0
[2016-10-14 21:26:03 - DEBUG2] Starting process: ["git", "--git-dir", "/vagrant/puppet/3rdparty/modules/solr/.git", "--work-tree", "/vagrant/puppet/3rdparty/modules/solr", "config", "--get", "remote.origin.url"]
[2016-10-14 21:26:03 - DEBUG2] Finished process:
Command: git --git-dir /vagrant/puppet/3rdparty/modules/solr/.git --work-tree /vagrant/puppet/3rdparty/modules/solr config --get remote.origin.url
Stdout:
https://bitbucket.org/hdeadman/puppet-solr.git
Exit code: 0
[2016-10-14 21:26:03 - DEBUG2] Starting process: ["git", "--git-dir", "/vagrant/puppet/3rdparty/modules/solr/.git", "--work-tree", "/vagrant/puppet/3rdparty/modules/solr", "diff-index", "--exit-code", "--name-only", "HEAD"]
[2016-10-14 21:26:03 - DEBUG2] Finished process:
Command: git --git-dir /vagrant/puppet/3rdparty/modules/solr/.git --work-tree /vagrant/puppet/3rdparty/modules/solr diff-index --exit-code --name-only HEAD
Stdout:
.gitignore
.travis.yml
CHANGELOG.md
Gemfile
LICENSE
README.md
Rakefile
contributors.txt
files/jetty-logging.xml
lib/puppet/parser/functions/other_files.rb
manifests/config.pp
manifests/core.pp
manifests/init.pp
manifests/install.pp
manifests/params.pp
manifests/service.pp
manifests/shared_lib.pp
metadata.json
spec/classes/init_spec.rb
spec/spec_helper.rb
templates/solr.in.sh.erb
templates/solr.service.erb
templates/solr.sh.erb
tests/init.pp
tests/landcare_test.pp
vagrant/Vagrantfile
vagrant/config/centos6.6.rb
vagrant/config/centos7.0.rb
vagrant/config/centos7.2.rb
vagrant/config/testos.rb
vagrant/config/ubuntu12.04.rb
vagrant/config/ubuntu14.04.rb
vagrant/config/ubuntu16.04.rb
vagrant/nbproject/project.properties
vagrant/nbproject/project.xml
vagrant/puppet/Puppetfile
vagrant/puppet/test/.blank
vagrant/scripts/noop.sh
vagrant/scripts/provision.bash
vagrant/scripts/redhat.sh
vagrant/scripts/run.bash
vagrant/scripts/ubuntu16.04.sh
vagrant/scripts/upgrade-puppet.sh
vagrant/test.bash
Exit code: 1
[2016-10-14 21:26:03 - DEBUG] Found local modifications in /vagrant/puppet/3rdparty/modules/solr/.gitignore
.travis.yml
CHANGELOG.md
Gemfile
LICENSE
README.md
Rakefile
contributors.txt
files/jetty-logging.xml
lib/puppet/parser/functions/other_files.rb
manifests/config.pp
manifests/core.pp
manifests/init.pp
manifests/install.pp
manifests/params.pp
manifests/service.pp
manifests/shared_lib.pp
metadata.json
spec/classes/init_spec.rb
spec/spec_helper.rb
templates/solr.in.sh.erb
templates/solr.service.erb
templates/solr.sh.erb
tests/init.pp
tests/landcare_test.pp
vagrant/Vagrantfile
vagrant/config/centos6.6.rb
vagrant/config/centos7.0.rb
vagrant/config/centos7.2.rb
vagrant/config/testos.rb
vagrant/config/ubuntu12.04.rb
vagrant/config/ubuntu14.04.rb
vagrant/config/ubuntu16.04.rb
vagrant/nbproject/project.properties
vagrant/nbproject/project.xml
vagrant/puppet/Puppetfile
vagrant/puppet/test/.blank
vagrant/scripts/noop.sh
vagrant/scripts/provision.bash
vagrant/scripts/redhat.sh
vagrant/scripts/run.bash
vagrant/scripts/ubuntu16.04.sh
vagrant/scripts/upgrade-puppet.sh
vagrant/test.bash
[2016-10-14 21:26:03 - DEBUG2] Starting process: ["git", "--git-dir", "/vagrant/puppet/3rdparty/modules/solr/.git", "--work-tree", "/vagrant/puppet/3rdparty/modules/solr", "diff-index", "-p", "HEAD", ".gitignore\n.travis.yml\nCHANGELOG.md\nGemfile\nLICENSE\nREADME.md\nRakefile\ncontributors.txt\nfiles/jetty-logging.xml\nlib/puppet/parser/functions/other_files.rb\nmanifests/config.pp\nmanifests/core.pp\nmanifests/init.pp\nmanifests/install.pp\nmanifests/params.pp\nmanifests/service.pp\nmanifests/shared_lib.pp\nmetadata.json\nspec/classes/init_spec.rb\nspec/spec_helper.rb\ntemplates/solr.in.sh.erb\ntemplates/solr.service.erb\ntemplates/solr.sh.erb\ntests/init.pp\ntests/landcare_test.pp\nvagrant/Vagrantfile\nvagrant/config/centos6.6.rb\nvagrant/config/centos7.0.rb\nvagrant/config/centos7.2.rb\nvagrant/config/testos.rb\nvagrant/config/ubuntu12.04.rb\nvagrant/config/ubuntu14.04.rb\nvagrant/config/ubuntu16.04.rb\nvagrant/nbproject/project.properties\nvagrant/nbproject/project.xml\nvagrant/puppet/Puppetfile\nvagrant/puppet/test/.blank\nvagrant/scripts/noop.sh\nvagrant/scripts/provision.bash\nvagrant/scripts/redhat.sh\nvagrant/scripts/run.bash\nvagrant/scripts/ubuntu16.04.sh\nvagrant/scripts/upgrade-puppet.sh\nvagrant/test.bash"]
[2016-10-14 21:26:03 - DEBUG2] Finished process:
Command: git --git-dir /vagrant/puppet/3rdparty/modules/solr/.git --work-tree /vagrant/puppet/3rdparty/modules/solr diff-index -p HEAD .gitignore
.travis.yml
CHANGELOG.md
Gemfile
LICENSE
README.md
Rakefile
contributors.txt
files/jetty-logging.xml
lib/puppet/parser/functions/other_files.rb
manifests/config.pp
manifests/core.pp
manifests/init.pp
manifests/install.pp
manifests/params.pp
manifests/service.pp
manifests/shared_lib.pp
metadata.json
spec/classes/init_spec.rb
spec/spec_helper.rb
templates/solr.in.sh.erb
templates/solr.service.erb
templates/solr.sh.erb
tests/init.pp
tests/landcare_test.pp
vagrant/Vagrantfile
vagrant/config/centos6.6.rb
vagrant/config/centos7.0.rb
vagrant/config/centos7.2.rb
vagrant/config/testos.rb
vagrant/config/ubuntu12.04.rb
vagrant/config/ubuntu14.04.rb
vagrant/config/ubuntu16.04.rb
vagrant/nbproject/project.properties
vagrant/nbproject/project.xml
vagrant/puppet/Puppetfile
vagrant/puppet/test/.blank
vagrant/scripts/noop.sh
vagrant/scripts/provision.bash
vagrant/scripts/redhat.sh
vagrant/scripts/run.bash
vagrant/scripts/ubuntu16.04.sh
vagrant/scripts/upgrade-puppet.sh
vagrant/test.bash
Stderr:
fatal: ambiguous argument '.gitignore
.travis.yml
CHANGELOG.md
Gemfile
LICENSE
README.md
Rakefile
contributors.txt
files/jetty-logging.xml
lib/puppet/parser/functions/other_files.rb
manifests/config.pp
manifests/core.pp
manifests/init.pp
manifests/install.pp
manifests/params.pp
manifests/service.pp
manifests/shared_lib.pp
metadata.json
spec/classes/init_spec.rb
spec/spec_helper.rb
templates/solr.in.sh.erb
templates/solr.service.erb
templates/solr.sh.erb
tests/init.pp
tests/landcare_test.pp
vagrant/Vagrantfile
vagrant/config/centos6.6.rb
vagrant/config/centos7.0.rb
vagrant/config/centos7.2.rb
vagrant/config/testos.rb
vagrant/config/ubuntu12.04.rb
vagrant/config/ubuntu14.04.rb
vagrant/config/ubuntu16.04.rb
vagrant/nbproject/project.properties
vagrant/nbproject/project.xml
vagrant/puppet/Puppetfile
vagrant/puppet/test/.blank
vagrant/scripts/noop.sh
vagrant/scripts/provision.bash
vagrant/scripts/redhat.sh
vagrant/scripts/run.bash
vagrant/scripts/ubuntu16.04.sh
vagrant/scripts/upgrade-puppet.sh
vagrant/test.bash': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Exit code: 128
[2016-10-14 21:26:03 - DEBUG1]
[2016-10-14 21:26:03 - WARN] Skipping /vagrant/puppet/3rdparty/modules/solr due to local modifications
```

After running r10k I ran this to show the module has no local modifications:

```
[root@chat1 3rdparty]# cd modules/solr
[root@chat1 solr]# git status
# HEAD detached at 03c429c
nothing to commit, working directory clean
[root@chat1 solr]#
```

Oddly, after doing that if I run r10k again, the modules where I have run "git status" no longer come up as being skipped. The output for that module looks like this on the subsequent run:

```
[2016-10-14 21:31:27 - INFO] Updating module /vagrant/puppet/3rdparty/modules/solr
[2016-10-14 21:31:27 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "rev-parse", "master^{commit}"]
[2016-10-14 21:31:27 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git rev-parse master^{commit}
Stdout:
03c429c2ba8305e75bb23b9e9b5f5bec057e90dd
Exit code: 0
[2016-10-14 21:31:27 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "for-each-ref", "refs/heads", "--format", "%(refname)"]
[2016-10-14 21:31:27 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git for-each-ref refs/heads --format %(refname)
Stdout:
refs/heads/master
Exit code: 0
[2016-10-14 21:31:27 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "config", "--local", "--get-regexp", "^remote\\..*\\.url$"]
[2016-10-14 21:31:27 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git config --local --get-regexp ^remote\..*\.url$
Stdout:
remote.origin.url https://bitbucket.org/hdeadman/puppet-solr.git
Exit code: 0
[2016-10-14 21:31:27 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "fetch", "origin", "--prune"]
[2016-10-14 21:31:28 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git fetch origin --prune
Exit code: 0
[2016-10-14 21:31:28 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "rev-parse", "master^{commit}"]
[2016-10-14 21:31:28 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git rev-parse master^{commit}
Stdout:
03c429c2ba8305e75bb23b9e9b5f5bec057e90dd
Exit code: 0
[2016-10-14 21:31:28 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "rev-parse", "master^{commit}"]
[2016-10-14 21:31:28 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git rev-parse master^{commit}
Stdout:
03c429c2ba8305e75bb23b9e9b5f5bec057e90dd
Exit code: 0
[2016-10-14 21:31:28 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "for-each-ref", "refs/heads", "--format", "%(refname)"]
[2016-10-14 21:31:28 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git for-each-ref refs/heads --format %(refname)
Stdout:
refs/heads/master
Exit code: 0
[2016-10-14 21:31:28 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "rev-parse", "master^{commit}"]
[2016-10-14 21:31:28 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git rev-parse master^{commit}
Stdout:
03c429c2ba8305e75bb23b9e9b5f5bec057e90dd
Exit code: 0
[2016-10-14 21:31:28 - DEBUG2] Starting process: ["git", "--git-dir", "/vagrant/puppet/3rdparty/modules/solr/.git", "--work-tree", "/vagrant/puppet/3rdparty/modules/solr", "config", "--get", "remote.origin.url"]
[2016-10-14 21:31:28 - DEBUG2] Finished process:
Command: git --git-dir /vagrant/puppet/3rdparty/modules/solr/.git --work-tree /vagrant/puppet/3rdparty/modules/solr config --get remote.origin.url
Stdout:
https://bitbucket.org/hdeadman/puppet-solr.git
Exit code: 0
[2016-10-14 21:31:28 - DEBUG2] Starting process: ["git", "--git-dir", "/vagrant/puppet/3rdparty/modules/solr/.git", "--work-tree", "/vagrant/puppet/3rdparty/modules/solr", "diff-index", "--exit-code", "--name-only", "HEAD"]
[2016-10-14 21:31:28 - DEBUG2] Finished process:
Command: git --git-dir /vagrant/puppet/3rdparty/modules/solr/.git --work-tree /vagrant/puppet/3rdparty/modules/solr diff-index --exit-code --name-only HEAD
Exit code: 0
[2016-10-14 21:31:28 - DEBUG2] Starting process: ["git", "--git-dir", "/vagrant/puppet/3rdparty/modules/solr/.git", "rev-parse", "HEAD^{commit}"]
[2016-10-14 21:31:28 - DEBUG2] Finished process:
Command: git --git-dir /vagrant/puppet/3rdparty/modules/solr/.git rev-parse HEAD^{commit}
Stdout:
03c429c2ba8305e75bb23b9e9b5f5bec057e90dd
Exit code: 0
[2016-10-14 21:31:28 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "rev-parse", "master^{commit}"]
[2016-10-14 21:31:28 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git rev-parse master^{commit}
Stdout:
03c429c2ba8305e75bb23b9e9b5f5bec057e90dd
Exit code: 0
[2016-10-14 21:31:28 - DEBUG2] Starting process: ["git", "--git-dir", "/root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git", "for-each-ref", "refs/heads", "--format", "%(refname)"]
[2016-10-14 21:31:28 - DEBUG2] Finished process:
Command: git --git-dir /root/.r10k/git/https---bitbucket.org-hdeadman-puppet-solr.git for-each-ref refs/heads --format %(refname)
Stdout:
refs/heads/master
Exit code: 0
[2016-10-14 21:31:28 - DEBUG] /vagrant/puppet/3rdparty/modules/solr is already at Git ref master

```

**Comment by fatmcgav** [comment]
Looks like @netdata raised #705, but it needs a rebase...

**Comment by ustuehler** [comment]
@scotje @hdeadman, the underlying issue is that `git diff-index` returns a exit code 1 until running `git status`. After that, `git diff-index` properly returns 0.

It looks like `git status` has the side-effect of initialising the index while `git diff-index` does not. This should be a workaround for `r10k puppetfile install` skipping git modules:

```console
$ for d in modules/*/.git; do (cd $d/.. && git status >/dev/null); done
```
